<a href="https://colab.research.google.com/github/RubenMcCarty/Desicion_Making_Model_UPN/blob/main/Lab_05_Adquisicion_Datos_Introducci%C3%B3n.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Introducción a Notebook-Jupyter -Adquisición de Datos de Dataset de coches**

# [MSc: Ruben Quispe](https://www.linkedin.com/in/msc-rub%C3%A9n-quispe-l/)

# Objetivos
Despues de completar este laboratorio podras saber:
* Adequerir datos de varias formas
* Obtener información de Pandas (Libreria de python)

# Table de contenido

1. Data Acquisition
2. Conocimiento básico de conjunto de datos (dataset)

# **Data Adquisition  o adquisicion de datos**

Hay varios formatos para un dataset .csv, .json, .xlsx, etc.

El dataset se puede almacenar en diferentes lugares, en su PC local o en linea.

En esta sesión, aprenderas cómo cargar dataset a un notebook jupyter.

En nuestro caso vamos a cargar datos de automovil que esta en linea en formato .csv ( commas separate value-Valores separados por comas)

Fuente de datos: https://archive.ics.uci.edu/ml/machine-learning-databases/autos/imports-85.data

Tipo de dato: .csv

La libreria pandas, nos permite trabajar con dataset, cargar dataframe, leer varios dataset, para debemos importa pandas



# 1 Importar las librerías necesarias

In [1]:
import pandas as pd # para trabajar con dataframe
import numpy as np  # para trabajar con matrices, operadores matemáticos

In [2]:
a = np.zeros((2, 3))
a

array([[0., 0., 0.],
       [0., 0., 0.]])

In [3]:
b = np.ones((4,5))
b

array([[1., 1., 1., 1., 1.],
       [1., 1., 1., 1., 1.],
       [1., 1., 1., 1., 1.],
       [1., 1., 1., 1., 1.]])

# 2 Leer nuestros datos
Usamos la función pandas.read_csv() para leer los datos del csv, en el parentesis colocamos la ruta del archivo entre comillas, la ruta puede ser un url o la dirección de su archivo local

Debido a que los datos no incluye encabezados podemos agregar un argumento encabezados.


In [4]:
dato = "https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-DA0101EN-SkillsNetwork/labs/Data%20files/auto.csv"

In [5]:
df = pd.read_csv(dato, header= None) 

In [6]:
df

0    1            2       3      4     5   ...    20   21    22  23  24     25
0     3    ?  alfa-romero     gas    std   two  ...   9.0  111  5000  21  27  13495
1     3    ?  alfa-romero     gas    std   two  ...   9.0  111  5000  21  27  16500
2     1    ?  alfa-romero     gas    std   two  ...   9.0  154  5000  19  26  16500
3     2  164         audi     gas    std  four  ...  10.0  102  5500  24  30  13950
4     2  164         audi     gas    std  four  ...   8.0  115  5500  18  22  17450
..   ..  ...          ...     ...    ...   ...  ...   ...  ...   ...  ..  ..    ...
200  -1   95        volvo     gas    std  four  ...   9.5  114  5400  23  28  16845
201  -1   95        volvo     gas  turbo  four  ...   8.7  160  5300  19  25  19045
202  -1   95        volvo     gas    std  four  ...   8.8  134  5500  18  23  21485
203  -1   95        volvo  diesel  turbo  four  ...  23.0  106  4800  26  27  22470
204  -1   95        volvo     gas  turbo  four  ...   9.5  114  5400  19  25  22625

[205 rows x 26 columns]

In [7]:
df.head()

0    1            2    3    4     5   ...    20   21    22  23  24     25
0   3    ?  alfa-romero  gas  std   two  ...   9.0  111  5000  21  27  13495
1   3    ?  alfa-romero  gas  std   two  ...   9.0  111  5000  21  27  16500
2   1    ?  alfa-romero  gas  std   two  ...   9.0  154  5000  19  26  16500
3   2  164         audi  gas  std  four  ...  10.0  102  5500  24  30  13950
4   2  164         audi  gas  std  four  ...   8.0  115  5500  18  22  17450

[5 rows x 26 columns]

In [8]:
df.tail()

0   1      2       3      4     5   ...    20   21    22  23  24     25
200  -1  95  volvo     gas    std  four  ...   9.5  114  5400  23  28  16845
201  -1  95  volvo     gas  turbo  four  ...   8.7  160  5300  19  25  19045
202  -1  95  volvo     gas    std  four  ...   8.8  134  5500  18  23  21485
203  -1  95  volvo  diesel  turbo  four  ...  23.0  106  4800  26  27  22470
204  -1  95  volvo     gas  turbo  four  ...   9.5  114  5400  19  25  22625

[5 rows x 26 columns]

In [9]:
df.shape # nos indica la cantidad de filas y columnas

(205, 26)

# Pregunta 1
Mostrar los 10 primeros filas del dataframe

In [10]:
print('Los 10 primeras filas del dataframe son\n')
df.head(10)

Los 10 primeras filas del dataframe son



0    1            2    3      4     5   ...    20   21    22  23  24     25
0   3    ?  alfa-romero  gas    std   two  ...   9.0  111  5000  21  27  13495
1   3    ?  alfa-romero  gas    std   two  ...   9.0  111  5000  21  27  16500
2   1    ?  alfa-romero  gas    std   two  ...   9.0  154  5000  19  26  16500
3   2  164         audi  gas    std  four  ...  10.0  102  5500  24  30  13950
4   2  164         audi  gas    std  four  ...   8.0  115  5500  18  22  17450
5   2    ?         audi  gas    std   two  ...   8.5  110  5500  19  25  15250
6   1  158         audi  gas    std  four  ...   8.5  110  5500  19  25  17710
7   1    ?         audi  gas    std  four  ...   8.5  110  5500  19  25  18920
8   1  158         audi  gas  turbo  four  ...   8.3  140  5500  17  20  23875
9   0    ?         audi  gas  turbo   two  ...   7.0  160  5500  16  22      ?

[10 rows x 26 columns]

In [11]:
print ('BIENVENIDO A PYTHON')

BIENVENIDO A PYTHON


In [12]:
print("MODELO PARA LA TOMA DE DECISIONES")

MODELO PARA LA TOMA DE DECISIONES


# Pregunta 2 
Imprima las 10 ultimas filas del dataframe

In [13]:
df.tail(10)

0    1      2       3      4     5   ...    20   21    22  23  24     25
195  -1   74  volvo     gas    std  four  ...   9.5  114  5400  23  28  13415
196  -2  103  volvo     gas    std  four  ...   9.5  114  5400  24  28  15985
197  -1   74  volvo     gas    std  four  ...   9.5  114  5400  24  28  16515
198  -2  103  volvo     gas  turbo  four  ...   7.5  162  5100  17  22  18420
199  -1   74  volvo     gas  turbo  four  ...   7.5  162  5100  17  22  18950
200  -1   95  volvo     gas    std  four  ...   9.5  114  5400  23  28  16845
201  -1   95  volvo     gas  turbo  four  ...   8.7  160  5300  19  25  19045
202  -1   95  volvo     gas    std  four  ...   8.8  134  5500  18  23  21485
203  -1   95  volvo  diesel  turbo  four  ...  23.0  106  4800  26  27  22470
204  -1   95  volvo     gas  turbo  four  ...   9.5  114  5400  19  25  22625

[10 rows x 26 columns]

# AGREGAR CABEZERA

In [14]:
headers = ["symboling","normalized-losses","make","fuel-type","aspiration", "num-of-doors","body-style",
         "drive-wheels","engine-location","wheel-base", "length","width","height","curb-weight","engine-type",
         "num-of-cylinders", "engine-size","fuel-system","bore","stroke","compression-ratio","horsepower",
         "peak-rpm","city-mpg","highway-mpg","price"]
print("headers\n", headers)

headers
 ['symboling', 'normalized-losses', 'make', 'fuel-type', 'aspiration', 'num-of-doors', 'body-style', 'drive-wheels', 'engine-location', 'wheel-base', 'length', 'width', 'height', 'curb-weight', 'engine-type', 'num-of-cylinders', 'engine-size', 'fuel-system', 'bore', 'stroke', 'compression-ratio', 'horsepower', 'peak-rpm', 'city-mpg', 'highway-mpg', 'price']


In [16]:
df.columns = headers
df.head()

symboling normalized-losses         make  ... city-mpg highway-mpg  price
0          3                 ?  alfa-romero  ...       21          27  13495
1          3                 ?  alfa-romero  ...       21          27  16500
2          1                 ?  alfa-romero  ...       19          26  16500
3          2               164         audi  ...       24          30  13950
4          2               164         audi  ...       18          22  17450

[5 rows x 26 columns]

In [18]:
df.tail(30)

symboling normalized-losses        make  ... city-mpg highway-mpg  price
175         -1                65      toyota  ...       27          32   9988
176         -1                65      toyota  ...       27          32  10898
177         -1                65      toyota  ...       27          32  11248
178          3               197      toyota  ...       20          24  16558
179          3               197      toyota  ...       19          24  15998
180         -1                90      toyota  ...       20          24  15690
181         -1                 ?      toyota  ...       19          24  15750
182          2               122  volkswagen  ...       37          46   7775
183          2               122  volkswagen  ...       27          34   7975
184          2                94  volkswagen  ...       37          46   7995
185          2                94  volkswagen  ...       27          34   8195
186          2                94  volkswagen  ...       27          34   8495
187          2                94  volkswagen  ...       37          42   9495
188          2                94  volkswagen  ...       26          32   9995
189          3                 ?  volkswagen  ...       24          29  11595
190          3               256  volkswagen  ...       24          29   9980
191          0                 ?  volkswagen  ...       19          24  13295
192          0                 ?  volkswagen  ...       33          38  13845
193          0                 ?  volkswagen  ...       25          31  12290
194         -2               103       volvo  ...       23          28  12940
195         -1                74       volvo  ...       23          28  13415
196         -2               103       volvo  ...       24          28  15985
197         -1                74       volvo  ...       24          28  16515
198         -2               103       volvo  ...       17          22  18420
199         -1                74       volvo  ...       17          22  18950
200         -1                95       volvo  ...       23          28  16845
201         -1                95       volvo  ...       19          25  19045
202         -1                95       volvo  ...       18          23  21485
203         -1                95       volvo  ...       26          27  22470
204         -1                95       volvo  ...       19          25  22625

[30 rows x 26 columns]

necesitamos reemplazar el "?" símbolo con NaN para que dropna () pueda eliminar los valores faltantes

In [19]:
df1 = df.replace ('?', np.NaN)

In [20]:
df1

symboling normalized-losses         make  ... city-mpg highway-mpg  price
0            3               NaN  alfa-romero  ...       21          27  13495
1            3               NaN  alfa-romero  ...       21          27  16500
2            1               NaN  alfa-romero  ...       19          26  16500
3            2               164         audi  ...       24          30  13950
4            2               164         audi  ...       18          22  17450
..         ...               ...          ...  ...      ...         ...    ...
200         -1                95        volvo  ...       23          28  16845
201         -1                95        volvo  ...       19          25  19045
202         -1                95        volvo  ...       18          23  21485
203         -1                95        volvo  ...       26          27  22470
204         -1                95        volvo  ...       19          25  22625

[205 rows x 26 columns]

Ahora, hemos leído correctamente el conjunto de datos sin procesar y agregamos los encabezados correctos en el marco de datos.

In [22]:
df1.head(10)

symboling normalized-losses         make  ... city-mpg highway-mpg  price
0          3               NaN  alfa-romero  ...       21          27  13495
1          3               NaN  alfa-romero  ...       21          27  16500
2          1               NaN  alfa-romero  ...       19          26  16500
3          2               164         audi  ...       24          30  13950
4          2               164         audi  ...       18          22  17450
5          2               NaN         audi  ...       19          25  15250
6          1               158         audi  ...       19          25  17710
7          1               NaN         audi  ...       19          25  18920
8          1               158         audi  ...       17          20  23875
9          0               NaN         audi  ...       16          22    NaN

[10 rows x 26 columns]

Podemos eliminar los valores que faltan a lo largo de la columna "precio" de la siguiente manera

In [26]:
df=df1.dropna(subset=["price"], axis=0)
df.head(20)

symboling normalized-losses         make  ... city-mpg highway-mpg  price
0           3               NaN  alfa-romero  ...       21          27  13495
1           3               NaN  alfa-romero  ...       21          27  16500
2           1               NaN  alfa-romero  ...       19          26  16500
3           2               164         audi  ...       24          30  13950
4           2               164         audi  ...       18          22  17450
5           2               NaN         audi  ...       19          25  15250
6           1               158         audi  ...       19          25  17710
7           1               NaN         audi  ...       19          25  18920
8           1               158         audi  ...       17          20  23875
10          2               192          bmw  ...       23          29  16430
11          0               192          bmw  ...       23          29  16925
12          0               188          bmw  ...       21          28  20970
13          0               188          bmw  ...       21          28  21105
14          1               NaN          bmw  ...       20          25  24565
15          0               NaN          bmw  ...       16          22  30760
16          0               NaN          bmw  ...       16          22  41315
17          0               NaN          bmw  ...       15          20  36880
18          2               121    chevrolet  ...       47          53   5151
19          1                98    chevrolet  ...       38          43   6295
20          0                81    chevrolet  ...       38          43   6575

[20 rows x 26 columns]

Ahora, hemos leído correctamente el conjunto de datos sin procesar y agregamos los encabezados correctos en el marco de datos.

# Quiz encuentre el nombre de las columnas del dataframe

In [ ]:
print(df.columns)

Index(['symboling', 'normalized-losses', 'make', 'fuel-type', 'aspiration',
       'num-of-doors', 'body-style', 'drive-wheels', 'engine-location',
       'wheel-base', 'length', 'width', 'height', 'curb-weight', 'engine-type',
       'num-of-cylinders', 'engine-size', 'fuel-system', 'bore', 'stroke',
       'compression-ratio', 'horsepower', 'peak-rpm', 'city-mpg',
       'highway-mpg', 'price'],
      dtype='object')


# Guardar Dataset
En consecuencia, Pandas nos permite guardar el conjunto de datos en csv utilizando el método dataframe.to_csv (), puede agregar la ruta y el nombre del archivo junto con las comillas entre paréntesis.

Por ejemplo, si guarda el marco de datos df como automóvil.csv en su máquina local, puede usar la sintaxis siguiente:
<h2>Read/Save Other Data Formats</h2>

| Data Formate |        Read       |            Save |
| ------------ | :---------------: | --------------: |
| csv          |  `pd.read_csv()`  |   `df.to_csv()` |
| json         |  `pd.read_json()` |  `df.to_json()` |
| excel        | `pd.read_excel()` | `df.to_excel()` |
| hdf          |  `pd.read_hdf()`  |   `df.to_hdf()` |
| sql          |  `pd.read_sql()`  |   `df.to_sql()` |
| ...          |        ...        |             ... |

In [27]:
df.to_csv('automovil.csv')

# Conocimiento básico del conjunto de datos
Después de leer los datos en el DF de Pandas, es hora de que exploremos el conjunto de datos.
Hay varias formas de obtener información esencial de los datos para ayudarnos a comprender mejor nuestro conjunto de datos.

# Tipos de datos
Los datos tienen una variedad de tipos.
Los principales tipos almacenados en los DF de Pandas son object, float, int, bool y datetime64. Para conocer mejor cada atributo, siempre es bueno que sepamos el tipo de datos de cada columna. En Pandas:

In [29]:
df.dtypes # me indica el tipo de datos de cada columna, devuelve una serie con el tipo de datos de cada columna.

symboling              int64
normalized-losses     object
make                  object
fuel-type             object
aspiration            object
num-of-doors          object
body-style            object
drive-wheels          object
engine-location       object
wheel-base           float64
length               float64
width                float64
height               float64
curb-weight            int64
engine-type           object
num-of-cylinders      object
engine-size            int64
fuel-system           object
bore                  object
stroke                object
compression-ratio    float64
horsepower            object
peak-rpm              object
city-mpg               int64
highway-mpg            int64
price                 object
dtype: object

devuelve una serie con el tipo de datos de cada columna.

In [30]:
print(df.dtypes)

symboling              int64
normalized-losses     object
make                  object
fuel-type             object
aspiration            object
num-of-doors          object
body-style            object
drive-wheels          object
engine-location       object
wheel-base           float64
length               float64
width                float64
height               float64
curb-weight            int64
engine-type           object
num-of-cylinders      object
engine-size            int64
fuel-system           object
bore                  object
stroke                object
compression-ratio    float64
horsepower            object
peak-rpm              object
city-mpg               int64
highway-mpg            int64
price                 object
dtype: object


Como resultado, como se muestra arriba, es claro ver que el tipo de datos de "symboling" y "curb-wieght" son int64, "pérdidas normalizadas" es objeto, y "base de ruedas" es float64, etc.

Estos tipos de datos se pueden cambiar; aprenderemos cómo lograr esto en un módulo posterior.

# Describe
Si quisiéramos obtener un resumen estadístico de cada columna, como el recuento, el valor medio de la columna, la desviación estándar de la columna, etc. Usamos el método de describe.

In [31]:
df.describe()

symboling  wheel-base  ...    city-mpg  highway-mpg
count  201.000000  201.000000  ...  201.000000   201.000000
mean     0.840796   98.797015  ...   25.179104    30.686567
std      1.254802    6.066366  ...    6.423220     6.815150
min     -2.000000   86.600000  ...   13.000000    16.000000
25%      0.000000   94.500000  ...   19.000000    25.000000
50%      1.000000   97.000000  ...   24.000000    30.000000
75%      2.000000  102.400000  ...   30.000000    34.000000
max      3.000000  120.900000  ...   49.000000    54.000000

[8 rows x 10 columns]

Esto muestra el resumen estadístico de todas las columnas de tipo numérico (int, float).
Por ejemplo, el atributo "simbolización" tiene 205 recuentos, el valor medio de esta columna es 0,83, la desviación estándar es 1,25, el valor mínimo es -2, el percentil 25 es 0, el percentil 50 es 1, el percentil 75 es 2 y el valor máximo es 3.
Sin embargo, ¿qué pasaría si también quisiéramos comprobar todas las columnas, incluidas las que son de tipo objeto?

NOTA: Este método proporcionará varias estadísticas de resumen, excluyendo los valores de NaN (No es un número).

Puede agregar un argumento include = "all" dentro del corchete. Vamos a intentarlo otra vez.

In [32]:
# Si desea leer todo el dato
df.describe(include = 'all')

symboling normalized-losses    make  ...    city-mpg highway-mpg  price
count   201.000000               164     201  ...  201.000000  201.000000    201
unique         NaN                51      22  ...         NaN         NaN    186
top            NaN               161  toyota  ...         NaN         NaN  13499
freq           NaN                11      32  ...         NaN         NaN      2
mean      0.840796               NaN     NaN  ...   25.179104   30.686567    NaN
std       1.254802               NaN     NaN  ...    6.423220    6.815150    NaN
min      -2.000000               NaN     NaN  ...   13.000000   16.000000    NaN
25%       0.000000               NaN     NaN  ...   19.000000   25.000000    NaN
50%       1.000000               NaN     NaN  ...   24.000000   30.000000    NaN
75%       2.000000               NaN     NaN  ...   30.000000   34.000000    NaN
max       3.000000               NaN     NaN  ...   49.000000   54.000000    NaN

[11 rows x 26 columns]

# Quiz 
Puede seleccionar las columnas de un marco de datos indicando el nombre de cada columna, por ejemplo, puede seleccionar las tres columnas de la siguiente manera:

In [33]:
df[['symboling', 'normalized-losses', 'make']]

symboling normalized-losses         make
0            3               NaN  alfa-romero
1            3               NaN  alfa-romero
2            1               NaN  alfa-romero
3            2               164         audi
4            2               164         audi
..         ...               ...          ...
200         -1                95        volvo
201         -1                95        volvo
202         -1                95        volvo
203         -1                95        volvo
204         -1                95        volvo

[201 rows x 3 columns]

# Quiz
Donde "columna" es el nombre de la columna, puede aplicar el método ".describe ()" para obtener las estadísticas de esas columnas de la siguiente manera:

In [35]:
df[['symboling', 'engine-size', 'price']].describe()

symboling  engine-size
count  201.000000   201.000000
mean     0.840796   126.875622
std      1.254802    41.546834
min     -2.000000    61.000000
25%      0.000000    98.000000
50%      1.000000   120.000000
75%      2.000000   141.000000
max      3.000000   326.000000

Ahora, proporciona el resumen estadístico de todas las columnas, incluidos los atributos de tipo objeto.
Ahora podemos ver cuántos valores únicos, cuál es el valor superior y la frecuencia del valor superior en las columnas con tipo de objeto.
Algunos valores en la tabla anterior se muestran como "NaN", esto se debe a que esos números no están disponibles con respecto a un tipo de columna en particular.

# info
Otro método que puede usar para verificar su conjunto de datos es:

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 201 entries, 0 to 204
Data columns (total 26 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   symboling          201 non-null    int64  
 1   normalized-losses  164 non-null    object 
 2   make               201 non-null    object 
 3   fuel-type          201 non-null    object 
 4   aspiration         201 non-null    object 
 5   num-of-doors       199 non-null    object 
 6   body-style         201 non-null    object 
 7   drive-wheels       201 non-null    object 
 8   engine-location    201 non-null    object 
 9   wheel-base         201 non-null    float64
 10  length             201 non-null    float64
 11  width              201 non-null    float64
 12  height             201 non-null    float64
 13  curb-weight        201 non-null    int64  
 14  engine-type        201 non-null    object 
 15  num-of-cylinders   201 non-null    object 
 16  engine-size        201 non

In [36]:
df2 = pd.read_csv('automovil.csv')
df2

Unnamed: 0  symboling  normalized-losses  ... city-mpg highway-mpg  price
0             0          3                NaN  ...       21          27  13495
1             1          3                NaN  ...       21          27  16500
2             2          1                NaN  ...       19          26  16500
3             3          2              164.0  ...       24          30  13950
4             4          2              164.0  ...       18          22  17450
..          ...        ...                ...  ...      ...         ...    ...
196         200         -1               95.0  ...       23          28  16845
197         201         -1               95.0  ...       19          25  19045
198         202         -1               95.0  ...       18          23  21485
199         203         -1               95.0  ...       26          27  22470
200         204         -1               95.0  ...       19          25  22625

[201 rows x 27 columns]